RecoTracker/DeDx/plugins/DeDxHitCalibrator.cc

```diff
@@ -0,0 +1,392 @@
+                                       reco::DeDxHitCollection& pixelHits,
+                                       reco::DeDxHitCollection& stripHits) {
+  for (size_t i = 0; i < info.size(); i++) {
+    // Require hits to be complete and compatible
+    const auto& type = info.type(i);
+    if (!(type & (1 << reco::DeDxHitInfo::Complete)) || !(type & (1 << reco::DeDxHitInfo::Compatible)))
+      continue;
+
+    // Effective path length
+    const auto& pl = info.pathlength(i);
+    const auto pathLength = pl * (1. + 0.07 * std::log(pl / 450e-4));
+    const DetId& detId = info.detId(i);
+
+    // Strip
+    if (const auto& stripCluster = info.stripCluster(i)) {
+      const auto& thickness =
+          dynamic_cast<const StripGeomDetUnit*>(tkGeom_->idToDet(detId))->surface().bounds().thickness();
+      const auto& det = getDetId(detId, thickness);
+      const auto& chipId = ChipId(detId, stripCluster->barycenter() / 128.);
+      // Fill measured strip deposits
+      const auto& thr = dedxCalib_->thr[det];
+      std::vector<std::pair<double, int> > b;
+      b.reserve(stripCluster->amplitudes().size() + 2);
+      b.emplace_back(thr, isBelow);
+      for (const auto& adc : stripCluster->amplitudes()) {
+        if (adc > 253)
+          b.emplace_back(254., isOver);
+        else
+          b.emplace_back(adc + 0.5, isNormal);
+      }
+      b.emplace_back(thr, isBelow);
+      // Fit
+      const auto& result = fitStripCluster(b, dedxCalib_->alpha[det], 1. / dedxCalib_->sigma[det]);
+      // ADC -> e- -> MeV
+      const auto& energy = correctEnergy(result.first * 217 * 3.61e-6, chipId);
+      const auto& esigma = correctEnergy(result.second * 217 * 3.61e-6, chipId);
+      // Compute charge
+      const auto& charge = pathLength != 0. ? energy / pathLength : 0.;
+      stripHits.emplace_back(charge, trackMomentum, pathLength, 0, esigma);
+    }
+
+    // Pixel
+    if (const auto& pixelCluster = info.pixelCluster(i)) {
+      const auto& thickness =
+          dynamic_cast<const PixelGeomDetUnit*>(tkGeom_->idToDet(detId))->surface().bounds().thickness();
+      const auto& det = getDetId(detId, thickness);
+      const auto& chipId = ChipId(detId, (int(pixelCluster->x() / 80.) << 3) + int(pixelCluster->y() / 52.));
+      // Collect adc
+      double delta(0);
+      bool isSaturated(false);
+      for (size_t j = 0; j < pixelCluster->pixelADC().size(); j++) {
+        const auto& elec = pixelCluster->pixelADC()[j];
+        delta += elec * 3.61e-6;
+        // ( pos , (x , y) )
+        const auto& row = pixelCluster->minPixelRow() + pixelCluster->pixelOffset()[2 * j];
+        const auto& col = pixelCluster->minPixelCol() + pixelCluster->pixelOffset()[2 * j + 1];
+        // Go back to adc
+        const auto& DBgain = pixelCalib_.getGain(detId, col, row);
+        const auto& DBpedestal = pixelCalib_.getPedestal(detId, col, row);
+        if (DBgain > 0.) {
+          const auto vcal = (elec + 414.) / 65.;
```

Comment:
This part is needed to go back from electrons to ADC, since we want information whether a given pixel/strip is in overflow. The values 414 and 65 were needed for this back-transformation. I agree, it should be substituted with something more reliable (I mean find the piece of code we want to reverse).

Reply:
I think what you want is the content of `SiPixelVCal` (see https://github.com/cms-sw/cmssw/pull/29829 for details).

Reply:
I tried to use SiPixelVCal but it seems it does not exist in any GlobalTag

Reply:
I did test it by running it on 2023 raw UPC PbPb data (it is shown in the slides attached in the PR description) but I will update to the latest conditions and rerun to check the impact

Reply:
In Run 3 we go straight from ADC->electrons (before it used to be a two-step process ADC->VCal->electrons). The SiPixelVCal payload is not included in any of the GlobalTags since it is not used directly in the reconstruction. It is only used in deriving the gain calibrations that are directly used in the reconstruction as explained in https://twiki.cern.ch/twiki/bin/view/CMS/SiPixelVCalHistory In fact, the content of the SiPixelVCal payload changes with time to account for irradiation effects in the detector. For simulation, on the other hand, the slopes and offsets are fixed to the values linked above in both the simulation and reconstruction. Phase 0 and Phase 1 values are of course different because these are two different detectors but even for Phase 1 L1 has separate value because it uses a readout chip different from the one used in other parts of the detector.

Reply:
If the target for this code is Run 3 data, the conversion constants should really come from the global tag through the use of the SiPixelGainCalibrationService. For MC one could in principle use hardcoded values because those stored in MC global tags are equal to the hardcoded values used in the digitizer.

Reply:
This looks ok and yes, you are right, for the offline reconstruction the SiPixelGainCalibrationOfflineService is used so that's the one your should use.

Reply:
@ferencek : I just checked using Run3 data that indeed the ADC value I get by undoing the pixel calibrations coincide with the ADC value of the pixel digi that has the same row, column and det id. However, there were a few cases were it did not coincide, in which the pixelADC()[j] returned always the same value of 65,535, which gives me the impression that the uint16_t parameter is not large enough to store the electron value (seems to be constrained to 65,535 despite that it should be larger if I multiply the ADC by the pixel gain value), is this expected or an issue?

Reply:
The value of 63535 is clear, this is the maximum value uint16_t can store (2^16-1). What I did not realize was that the saturation with the calibrated values in electrons could occur sooner than with raw ADC counts. I am not sure if this is universally true for all detIds since it depends on the exact values of calibration constants so what you propose, checking if either the ADC is 255 or the electron value is 65535, is a reasonable strategy.

Reply:
Would it be feasible to change from the data format uint16_t to uint32_t in the reconstruction code? or is uint16_t a constrained imposed by the detector?

Reply:
Added the check of electron == 65535 for pixel saturation in c65a578ddd929b9bad643a6ac3f315686db916a8